### ISSUE TITLE
`nix develop` relies on external nixpkgs even for current repo being nixpkgs

### ISSUE THREAD

**Issue by trofi** [NixOS/nix#5442]
**Describe the bug**

The simplest reproducer is to run `nix develop` in nixpkgs:

```
nixpkgs $ nix develop -f. re2c --no-use-registries
error (ignored): error: 'flake:nixpkgs' is an indirect flake reference, but registry lookups are not allowed
...
```

Ideally I would expect `nix develop` to use local `bash` and not download external `nixpkgs` copy.
Currently I work around redownload with `nix registry pin nixpkgs`. Ideally it would not be needed for nixpkgs `itself`.

**Expected behavior**

No external `nixpkgs` should be used.

**`nix-env --version` output**
nix-env (Nix) 2.5pre20211007_844dd90

**Additional context**

From what I understand flake reference comes from `src/nix/develop.cc`:

```c++
        Path shell = "bash";

        try {
            auto state = getEvalState();

            auto nixpkgsLockFlags = lockFlags;
            nixpkgsLockFlags.inputOverrides = {};
            nixpkgsLockFlags.inputUpdates = {};

            auto bashInstallable = std::make_shared<InstallableFlake>(
                this,
                state,
                installable->nixpkgsFlakeRef(),
                Strings{"bashInteractive"},
                Strings{"legacyPackages." + settings.thisSystem.get() + "."},
                nixpkgsLockFlags);

            shell = store->printStorePath(
                toStorePath(getEvalStore(), store, Realise::Outputs, OperateOn::Output, bashInstallable)) + "/bin/bash";
        } catch (Error &) {
            ignoreException();
        }
```

/cc @edolstra

**Comment by balsoft** [comment]
I would find this quite unexpected TBH. Why would the actual shell executable differ based on the flake we're trying to evaluate?

I think a better solution is to fall back to the system bash when we can't access nixpkgs.

**Comment by trofi** [comment]
I find the need to fetch some other version of `nixpkgs` every hour unexpected when I run `nix develop -f. foo` against local checkout of `nixpkgs` repository. I don't knowingly use flakes.

I personally work refetches around with `nix registry pin nixpkgs` currently (which matches neither my `bashInteractive` checkout nor my systems' `bashInteractive` install). It do it just to avoid 20-second initial startup delay. Is it really the way `nix develop` should be used? That sounds quite non-hermetic and slow.

**Comment by szlend** [comment]
I find it super frustrating that `nix develop` downloads a new version of `nixpkgs`/`bashInteractive` every time the nix registry cache expires. Having to wait minutes for the devshell to start makes the whole thing completely inconvenient to use.  Pinning  the `nixpkgs` registry works if you know what you're doing, but I cannot recommend this to colleagues who are unfamiliar with nix and will just forget they every pinned their registry and get stuck on some unmaintained revision forever.

> I would find this quite unexpected TBH. Why would the actual shell executable differ based on the flake we're trying to evaluate?

This is already the case in practice, just wait an hour or so. I feel like the current default is the worst of all worlds. The shell executable changes all the time, but without any caching benefits you would have if the version was locked in the flake.

Some ideas:
- Bundle `bashInteractive` with `nix`. Looking at `nix-tree nixpkgs#nix` output, it looks like `nix` already indirectly depends on `bash`.
- Default to using bash in `PATH` and bundle `bashInteractive` as a fallback when the system version is too old.
- Have a way to specify the devshell binary in the flake itself.

**Comment by trofi** [comment]
Yeah, having being bitten by slight desynch between `master` and unstable channel I ended up clobbering `nixpks` registry to always alias to the system's one in `flake.nix`:

```nix
          # avoid default registry and alias `nixpkgs` to the system.
          ({ ... }: {
            nix.registry = { nixpkgs.flake = nixpkgs; };
          })
```

**Comment by trofi** [comment]
Declaring it to be "works as intended".
